Fix the issue where the charset from Content-Type is not parsed correctly, and refactor the related code.

RainYuY opened the pull request:
## What is the purpose of the change?
Fix the issue where the charset from Content-Type is not parsed correctly, and refactor the related code.
Ref:issue #15187

## Checklist
- [x] Make sure there is a [GitHub_issue](https://github.com/apache/dubbo/issues) field for the change.
- [x] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
- [x] Write necessary unit-test to verify your logic correction. If the new feature or significant change is committed, please remember to add sample in [dubbo samples](https://github.com/apache/dubbo-samples) project.
- [x] Make sure gitHub actions can pass. [Why the workflow is failing and how to fix it?](../CONTRIBUTING.md)

RainYuY committed:
Fix the issue where the charset from Content-Type is not parsed correctly, and refactor the related code.

RainYuY commented:
@AlbumenJ @oxsean PTAL

kimmking commented on dubbo-remoting/dubbo-remoting-http12/src/test/java/org/apache/dubbo/remoting/http12/message/codec/HttpUtilsTest.java:
```diff
@@ -0,0 +1,35 @@
+        charset = HttpUtils.getCharsetFromContentType("text/html");
+        Assertions.assertEquals("", charset);
+        charset = HttpUtils.getCharsetFromContentType("application/json;charset=utf-8; boundary=__X_PAW_BOUNDARY__");
+        Assertions.assertEquals("utf-8", charset);
```
should be constructing a multipart request to ensure it works well as expected.

RainYuY replied:
I have added a multipart request. If you mean a completed HTTP connection test case, I think it would be better to add it to IT. I will add it after this PR is merged. Thanks for your CR.

brookqin commented:
@Stellar1999 
First of all, thank you for responding to this issue so promptly.

[rfc7231/3.1.1.1.  Media Type](https://datatracker.ietf.org/doc/html/rfc7231#section-3.1.1.1)
`media-type = type "/" subtype *( OWS ";" OWS parameter )`

According to the description in the RFC document, the `content-type` may include one or more parameters in addition to the media-type, with `charset` being just one of these parameters. There is no stipulated order for these parameters, so in extreme cases, relying on index positioning could potentially lead to issues.

Perhaps it is also necessary to define a new method: `String mediaTypeParam(String name);`, to retrieve the value of a specified parameter.
```
String mediaTypeParam(String name) {
  paramsStr = contentType.substring(X);
  params = paramsStr.split(";")
                    .filter(str -> str.contain("="))
                    .map(str -> str.split("="));
  return params[name];
}

var charset = mediaTypeParam("charset");
var boundary = mediaTypeParam("boundary");
```

RainYuY commented:
> @Stellar1999 First of all, thank you for responding to this issue so promptly.
> 
> [rfc7231/3.1.1.1. Media Type](https://datatracker.ietf.org/doc/html/rfc7231#section-3.1.1.1) `media-type = type "/" subtype *( OWS ";" OWS parameter )`
> 
> According to the description in the RFC document, the `content-type` may include one or more parameters in addition to the media-type, with `charset` being just one of these parameters. There is no stipulated order for these parameters, so in extreme cases, relying on index positioning could potentially lead to issues.
> 
> Perhaps it is also necessary to define a new method: `String mediaTypeParam(String name);`, to retrieve the value of a specified parameter.
> 
> ```
> String mediaTypeParam(String name) {
>   paramsStr = contentType.substring(X);
>   params = paramsStr.split(";")
>                     .filter(str -> str.contain("="))
>                     .map(str -> str.split("="));
>   return params[name];
> }
> 
> var charset = mediaTypeParam("charset");
> var boundary = mediaTypeParam("boundary");
> ```

Thanks for your review.
I think my new code does not rely on an index for positioning. (If you mean indexOf, it's only used to get the charset string.) Like you said, maybe we can find a better general method to retrieve more parameters. However, as I see in the Dubbo code, it only relies on the charset. So, I believe we just need a special method to obtain the charset.

oxsean commented on dubbo-remoting/dubbo-remoting-http12/src/main/java/org/apache/dubbo/remoting/http12/HttpUtils.java:
```diff
@@ -81,6 +81,22 @@ public static List<HttpCookie> decodeCookies(String value) {
+                charset = StringUtils.EMPTY_STRING;
+            } else {
+                charset = contentType.substring(index + CHARSET_PREFIX.length()).trim();
+                charset = charset.split(";")[0];
```
A tip, do not use split as it scans and slices the whole string and checks for regexp, use indexOf(char) for best performance.

RainYuY replied:
OK, This PR has be merged. I will submit another PR to do this fix.
Thanks for your review.

oxsean commented on dubbo-remoting/dubbo-remoting-http12/src/main/java/org/apache/dubbo/remoting/http12/HttpUtils.java:
```diff
@@ -81,6 +81,22 @@ public static List<HttpCookie> decodeCookies(String value) {
         return cookies;
     }
 
+    public static String getCharsetFromContentType(String contentType) {
```
parse or resolve will be btter than get.
parseCharset(String contentType)
Clear enough.
